spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/cloudfoundry/reactive/ReactiveTokenValidator.java

```diff
@@ -91,8 +92,7 @@
 	}
 
 	private void cacheTokenKeys(Map<String, String> tokenKeys) {
-		this.cachedTokenKeys.clear();
-		this.cachedTokenKeys.putAll(tokenKeys);
+		this.cachedTokenKeys = new HashMap<>(tokenKeys);
```

Comment:
I think this may need to be a `ConcurrentHashMap`. AFAIK, the field being volatile doesn't provide a happens-before relationship between the entries being added to the new map in the constructor and any subsequent reads or writes. That aside, I think it would be safer for it to be a `ConcurrentHashMap` anyway. If the code evolves such that the map is mutated, there would be a happens-before relationship between that mutation and any subsequent reads or writes.

WDYT, @izeye?

Reply:
@wilkinsona If I understand it correctly, as the current status, it's (effectively) immutable (could be decorated with unmodifiable wrapper to be explicit), so I think there's no visibility issue but if you think there's any chance to be mutable, I don't have a strong opinion `HashMap`. Let me know your final call.

Reply:
`new HashMap<>(tokenKeys)` mutates the map prior to returning it. There's no locking in `HashMap` so I don't think there's any guarantee that the mutation that's done in the constructor will be visible to another thread. I think we should use `ConcurrentHashMap`.

Reply:
@wilkinsona Sorry, I didn't understand your previous comment fully. With the additional comment, now I got your point 👍 I'll update this PR as you suggested.